pkg/controllers/job/plugins/distributed-framework/pytorch/pytorch.go

```diff
@@ -0,0 +1,188 @@
+	if masterIndex == -1 {
+		klog.Errorf("job %v doesn't have task %v", job.Name, pp.masterName)
+		return nil
+	}
+
+	masterEnvVars := []v1.EnvVar{}
+	masterAddr := pp.generateMasterAddr(job.Spec.Tasks[masterIndex], job.Name)
+	masterEnvVars = append(masterEnvVars, v1.EnvVar{
+		Name:  EnvMasterAddr,
+		Value: masterAddr,
+	}, v1.EnvVar{
+		Name:  EnvMasterPort,
+		Value: fmt.Sprintf("%v", pp.port),
+	})
+
+	masterRank := 0
+	workerRank := 0
+	if taskType == pp.workerName {
+		index, err := strconv.Atoi(helpers.GetPodIndexUnderTask(pod))
+		if err != nil {
+			return err
+		}
+
+		workerRank = index + 1
+	}
+
+	totalReplicas := pp.getTotalReplicas(job)
+	for i, c := range pod.Spec.Containers {
+		pp.openContainerPort(&c, i, pod)
+
+		pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, masterEnvVars...)
+		pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{
+			Name:  EnvWorldSize,
+			Value: strconv.Itoa(int(totalReplicas)),
+		})
+
+		if taskType == pp.workerName {
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{
+				Name:  EnvRank,
+				Value: strconv.Itoa(workerRank),
+			})
+		} else if taskType == pp.masterName {
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{
+				Name:  EnvRank,
+				Value: strconv.Itoa(masterRank),
+			})
+		}
+	}
+
+	return nil
+}
+
+func (pp *pytorchPlugin) getTotalReplicas(job *batch.Job) int32 {
+	jobReplicas := int32(0)
+	for _, task := range job.Spec.Tasks {
+		jobReplicas += task.Replicas
+	}
+
+	return jobReplicas
+}
+
+func (pp *pytorchPlugin) generateMasterAddr(task batch.TaskSpec, jobName string) string {
+	hostName := task.Template.Spec.Hostname
+	subdomain := task.Template.Spec.Subdomain
+	if len(hostName) == 0 {
+		hostName = helpers.MakePodName(jobName, task.Name, 0)
+	}
+	if len(subdomain) == 0 {
+		subdomain = jobName
+	}
+
+	host := hostName + "." + subdomain
+	return host
+}
+
+func (pp *pytorchPlugin) openContainerPort(c *v1.Container, index int, pod *v1.Pod) {
+	hasPort := false
+	for _, p := range c.Ports {
+		if p.ContainerPort == int32(pp.port) {
+			hasPort = true
+			break
+		}
+	}
+
+	if !hasPort {
+		port := v1.ContainerPort{
+			Name:          "pytorchjob-port",
+			ContainerPort: int32(pp.port),
+		}
+
+		pod.Spec.Containers[index].Ports = append(pod.Spec.Containers[index].Ports, port)
+	}
+}
+
+func (pp *pytorchPlugin) OnJobAdd(job *batch.Job) error {
+	if job.Status.ControlledResources["plugin-"+pp.Name()] == pp.Name() {
+		return nil
+	}
+	job.Status.ControlledResources["plugin-"+pp.Name()] = pp.Name()
+	return nil
+}
+
+func (pp *pytorchPlugin) OnJobDelete(job *batch.Job) error {
+	if job.Status.ControlledResources["plugin-"+pp.Name()] != pp.Name() {
+		return nil
+	}
+	delete(job.Status.ControlledResources, "plugin-"+pp.Name())
+	return nil
+}
+
+func (pp *pytorchPlugin) OnJobUpdate(job *batch.Job) error {
+	return nil
+}
+
+func (pp *pytorchPlugin) GetMasterName() string {
```

Comment:
Why this function is public?

Reply:
acutually it is not used, l can delete it